BadHabbit/BadHabbit/Kit/NetworkKit/RequestExecutor.swift

```diff
@@ -27,16 +28,37 @@ class RequestExecutor: RequestExecutorProtocol {
         self.config = config
     }
     
+    private func url(for request: Request) -> URL? {
+        var urlComponents = request.urlComponents()
+        urlComponents.scheme = config.scheme
+        urlComponents.host = config.host
+        
+        return urlComponents.url
+    }
+    
     func perform<Parser: ResponseParserProtocol>(
         request: Request,
         parser: Parser,
-        completion: (Result<Parser.Response, Error>) -> Void
+        completion: @escaping (Result<Parser.Response, Error>) -> Void
     ) {
-        do {
-            let response = try parser.parse(Data())
-            completion(.success(response))
-        } catch {
-            completion(.failure(error))
+        guard let url = url(for: request)
+        else {
+            completion(.failure(NetworkError.urlError))
+            return
         }
+        
+        URLSession.shared.dataTask(with: url) { data, _, error in
+            guard let data = data else {
```

Comment:
what if data is nil but error is still presented ?

Reply:
Does it better to return `error` passed into this closure or generic `dataError` ?

Reply:
I assumed here: _no data and no error_, but my code was wrong

Resolved

Reply:
Bug fixed (keyword return added)

Resolved

Reply:
In the latest update, `if error = nil && response.statuscode < 400`, the data will always have a value.
In our implementation, `if data.isEmpty {...}` we will throw a NetworkError.parseError

Resolved